src/service.js

```diff
@@ -586,22 +605,23 @@ module.exports = function(mixinOptions) {
 					this.apolloServer = new ApolloServer({
 						schema,
 						..._.defaultsDeep({}, mixinOptions.serverOptions, {
-							context: ({ req, connection }) => {
-								return req
-									? {
-											ctx: req.$ctx,
-											service: req.$service,
-											params: req.$params,
-											dataLoaders: new Map(), // create an empty map to load DataLoader instances into
-									  }
-									: {
-											service: connection.$service,
-									  };
-							},
+							context: ({ req, connection }) => ({
+								...(req ? {
+									ctx: req.$ctx,
+									service: req.$service,
+									params: req.$params,
+								} : {
+									ctx: connection.context.socket.$ctx,
+									service: connection.context.$service,
+									params: connection.context.socket.$params,
+								}),
+								dataLoaders: new Map(), // create an empty map to load DataLoader instances into
+							}),
 							subscriptions: {
-								onConnect: connectionParams => ({
-									...connectionParams,
-									$service: this,
+								onConnect: async (connectionParams, socket) => ({
+									...await this.actions.ws({ connectionParams, socket }),
```

Comment:
My instinct here is that there really isn't much of an advantage to mutating `socket` since the end result of all of this is to just add the resulting object on to context.  I feel like the `ws` action parameter signature is good as it will afford a lot of flexibility.  However, it seems like we could just return the full object that will become `context` from there, rather than mutating `socket` and using it.

That is, return an object with `ctx`, `service`, and `params` on it from the action and make that the return value from `onConnect`.  Is there an advantage to this approach that maybe I am not immediately seeing?

Reply:
The idea was to keep the same concept as moleculer-web (Where `req` & `res` are populated with `$ctx`). If anybody need to access the molculer context outside (Let's say some kind of magic ws middleware).
But your right, there is no other direct advantages other than keeping the moleculer-web approch, we can make the return value of `onContext` the action return value.
To me both approch seem good, what to you think ?

Reply:
Sorry, I'm just getting back to this now.  I do feel like the side-effecty behavior of the `ws` action mutating `socket` lends to a less understandable structure for it.  I think it would definitely be preferable to simply have it construct all of the `$xyz` variables that would ultimately become the `connection.context` object.

So this is what I'm thinking for the API on the `ws` action:

Input Params:
* `connectionParams`
* `socket`

Output Object:
```js
{
  $ctx: <Moleculer context>
  $socket: <Original socket parameter>
  $service: this,
  $params: <see more below>
}
```

With this structure, I feel like a consumer could wrap the `ws` action with hooks or middlewares and get access to everything provided on the way in (in a before-hook scenario) and also access a consistent payload of everything available on the way out (in an after-hook scenario).  My inclination is this might also put a path forward to alllowing consumers to solve for `authentication` issues by wrapping this action in a hook.  In our environment today we use `moleculer-web`'s `authentication` method to validate a JWT and then enrich `ctx.meta` with information.  I feel like a consumer here could wrap the `ws` method and implement whatever they wanted to for the websocket connections.

Back to `$params` for a minute.  I think it's a good idea, for consistency purposes, to provide a `$params` so it remains in a similar format to what happens on a normal request.  I do question whether `body` makes sense here though.  From what I'm understanding, `connectionParams` is more akin to request headers (please correct me if I'm wrong) than it would be to a body.  I'm not sure if it naturally is more like a `body` or `headers` but at least when I was playing in the GraphQL playground it seemed that I was able to specify this information in the `headers` input area.  I do see that `socket.upgradeReq` has its own `headers` property though, so there's a pretty good chance I'm off base here.

With respect to `socket.upgradeReq.query`, I'm not seeing that as a defined property on `socket.upgradeReq`.  Is there a scenario where this can be populated?

Reply:
Hey, sry for late reply,
After typing the code you're definitely right, this is more clean than mutating the socket, just one place where `$xyz` are built.

`$params` case
`connectionParams` should be based on the first message sent by the gql/web-socket client.
![image](https://user-images.githubusercontent.com/1301995/83773259-9aba0e80-a684-11ea-9826-c8497a660296.png)
This is why I thought of a "body", but this can also by considered as headers in a sense where this is send only at the beginning of the ws session, as you want for the final implementation.

For `query`, this one is really for specific use cases where you want to send data before the `connection_init` message.
A non related example is with "SignalR", the client send a bunch of information in the query (Excluding the auth token) : 
![image](https://camo.githubusercontent.com/c64e570987e7726968abffb160ad2459589144e2/68747470733a2f2f692e737461636b2e696d6775722e636f6d2f474d5530542e706e67)